### ISSUE TITLE
Helper scripts don't support cloud mode.

### ISSUE THREAD

**Issue by PaigingAlyssa** [docker-solr/docker-solr#183]
It would be nice to allow the various init scripts to support cloud mode. For example solr-create will only start up in local mode, create a core, and then restart in local mode. It would be nice if it had the option to start in cloud mode, create a collection, then restart into cloud mode.

**Comment by makuk66** [comment]
Can you be more specific about your use-case? What are you trying to achieve? How many nodes? How do you start your container(s)?

The cloud mode as documented in https://lucene.apache.org/solr/guide/7_4/getting-started-with-solrcloud.html#getting-started-with-solrcloud is geared at running multiple nodes on a single machine, whereas for Docker you're more likely to want to run multiple containers. That then suggests using Docker compose is more appropriate, but then the collection creation and dependencies are a bit more difficult to arrange. Coming up with some better up-to-date examples and/or scripts for that would definitely be a good thing.

**Comment by dsmiley** [comment]
So the two service/containers there "init-solr-home" and "solr" share data by using a volume mapping to the host machine to the same location "/docker/volumes/solr/solr_data"?  My docker is getting rusty so I just want to confirm what's going on here.

**Comment by PaigingAlyssa** [comment]
@dsmiley You are correct. The init-solr-home will start before the solr container because of the depends_on directive. Also init-solr-home exits after it sets up the solr_home directory, or if it already contains content, so it is expected to terminate fairly quickly.

**Comment by makuk66** [comment]
@PaigingAlyssa Have a look at https://github.com/docker-solr/docker-solr/pull/186:
```
git clone https://github.com/docker-solr/docker-solr.git
cd docker-solr
git checkout issue183-create-cloud-arg
./tools/build.sh 7.4/
docker run -p 8983:8983 --rm -it  docker-solr/docker-solr:7.4.0 solr-create -cloud -c mycollection
```

I still think doing it separately may be a better way forward; I'll experiment with that some too.

**Comment by PaigingAlyssa** [comment]
@makuk66 what does your build environment look like, these scripts are not Mac compatible :-(

**Comment by PaigingAlyssa** [comment]
@makuk66 I haven't had a chance to thoroughly test this, but it does seem to do what I requested at first pass. 

I still had some issues building in MacOS. The awk and readlink commands in build.sh don't appear to be compatible. MacOS doesn't have readlink, but homebrew's core utils does provide a greadlink. The awk --field-separator doesn't work either, but modifying to the -F option works. I made a few changes to those and got it building fine.

I would love to give a more through code review, but I am not a bash script expert, nor do I see any style guides to really adhere to. I will say it seems like a maintenance nightmare to have a mostly identical set of scripts duplicated across the entire project. I'm unsure if there is some kind of templating library/tool that could be added to avoid that.

Of course I am going to ask for more :-). What would this like to support clustering solr nodes? The biggest obstacle that I can see is the way zookeeper is being bootstrapped. That should probably become another feature request, but just throwing out thoughts.

Finally, thank you for taking the time to implement this.

**Comment by makuk66** [comment]
> I still had some issues building in MacOS.

Did you see my previous comment? Building on MacOS is supported by running the build itself inside a Docker container; that then builds without any modification being required. It's easier than trying to achieve full cross-platform compatibility.

> I would love to give a more through code review, but I am not a bash script expert, nor do I see any style guides to really adhere to.

No worries. My recommendation for bash is to run out [shellcheck](https://www.shellcheck.net) -- that's very helpful in pointing out obvious or potential issues, even if it can be somewhat overzealous at times. I spent some time this weekend making the docker-solr scripts compliant with the latest version of shellcheck. You can install it on MacOS with homebrew.

> I will say it seems like a maintenance nightmare to have a mostly identical set of scripts duplicated across the entire project. I'm unsure if there is some kind of templating library/tool that could be added to avoid that.

It is that way because that's how the library project automated builds create the various versions. We already have a templating solution: The Dockerfile templates (eg [Dockerfile.template](https://github.com/docker-solr/docker-solr/blob/master/Dockerfile.template)) live at the top-level, and the top-level [scripts](https://github.com/docker-solr/docker-solr/tree/master/scripts) directory contains the scripts. Running the [tools/update.sh](https://github.com/docker-solr/docker-solr/blob/master/tools/update.sh) helper script then does the template replacements and copies all the scripts to all the other directories. So basically, on MacOS, the development cycle looks like:

```
# make edits
vi scripts/*
# build one variant
(cd builder; ./build.sh; ./run.sh ./tools/build.sh 7.5/)
# test the new image with one relevant testcase; or test manually
(cd tests/create_core; ./test.sh docker-solr/docker-solr:7.5.0)
# update everything
tools/update.sh
# build and test everything
(cd builder; ./build.sh; ./run.sh ./tools/build_test_all.sh)
```

> Of course I am going to ask for more :-). What would this like to support clustering solr nodes? The biggest obstacle that I can see is the way zookeeper is being bootstrapped. That should probably become another feature request, but just throwing out thoughts.

Of course, and that's why I haven't merged this exploratory PR: if we're going to have to do something different to support further usecases, then perhaps this is not the way forward. Further work required.
